jmclient/jmclient/wallet_service.py

```diff
@@ -432,6 +434,25 @@ def get_address_usages(self):
         self.rewind_wallet_indices(used_indices, saved_indices)
         self.synced = True
 
+    def display_rescan_message_and_system_exit(self, restart_cb):
+        #TODO using system exit should be avoided as it makes the code
+        # harder to understand and reason about
+        #the whole idea of a restart callback is only needed because of exit
```

Comment:
I think maybe I need a little clarification to get your precise intent in this comment: are you saying (1) we should be able to fix it in future so restarts are not needed or (2) we shouldn't need restart_callbacks and (because?) we should throw Exceptions instead of calling a sys.exit()?

I don't know about (1), certainly we'd hope to do that, but about (2) I'm confused; I think at least for technical reasons it probably does make sense to go with a different Exception thrown rather than sys.exit() - e.g. see a recent issue #458 for some mess around this which I have no idea how to solve right now - but it also clearly (to me) is very logical to have a restart callback function since we have multiple types of clients that must be informed (currently only a CLI and a Qt GUI, sure, but could be more). Also there was a lot of mess around how restart callbacks were being propagated which iirc is cleaned up after #359 and a few related fixes.

Your point about sys.exit() being called in 2 different places is very valid.

Not complaining about this comment, more is always better as far as I'm concerned; I just wasn't 100% clear about it.

Reply:
My intent is (2). I disagree that callbacks are the most logical. Callbacks are not the only way to handle anomalous or exceptional conditions. Those can also be handled using the try: except: syntax, which I would argue is a better way for us partly because it's more readable as it puts error-handling code nearby in the same function as normal code. By constrast callbacks put the error handling code in another function which could be far away in the source file. So I'm imagining that in any script there would be a block of code `try: except WalletNeedsRescan: ` where within the `try` case is normal code about sync'ing the wallet while in the `except` case is code that's currently in rescan_cb. When used by a CLI script the `except` block can include `sys.exit()` but in all other cases it shouldn't. I'm not a big fan of `sys.exit()` anyway, as if it makes its way into any GUI code it makes the entire GUI disappear without any error message displayed to the user, which can be frustrating for them.

Reply:
I've always seen it as: the reason to use callbacks is to allow "injection": some base functionality doesn't need to know about how its events are handled by clients of it. To me that's orthogonal to whether you have a try: except: structure, or nearby-ness or other aspects of code structure (or indeed how errors are handled, with sys.exit or otherwise; and indeed the point you make at the end is exactly why there's a restart callback for the GUI).

The alternative to me is very inelegant: you have some base function (in this case a WalletService) that has to iterate over every type of client that calls it, in order to decide what to do in case of an event. Like here event = 'wallet needs a rescan' and you'd need effectively 'if caller is of this type do X, if caller is of this type do Y, ...' etc.
With callbacks the base function needs zero changes when a new type of client is created.

So anyway agreed sys.exit is a problematic/bad way of handling the restart requirement, but restart_callback (and callbacks generally) still seems to me like the right way of doing things, especially for this kind of software.

Reply:
I understand, I think I was still thinking in terms of the pre-wallet-service architecture. You're right that raising exceptions would have to be handled within wallet_service and not within any CLI or GUI scripts. We already heavily use twisted and that makes heavy use of callbacks.